rfcs/45-stageless.md

```diff
@@ -0,0 +1,584 @@
+As we add more of these atomic constraints (particularly if they share a single ancestor), the risk of these problems grows.
+In effect, atomically-connected systems look "almost like" a single system from the scheduler's perspective.
+
+On the other hand, atomic ordering constraints can be helpful when attempting to split large complex systems into multiple parts.
+By guaranteeing that the initial state of your complex system cannot be altered before the later parts of the system are complete,
+you can safely parallelize the rest of the work into seperate systems, improving both performance and maintainability.
+
+### States
+
+**States** are one particularly common and powerful form of run criteria, allowing you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three sets of systems:
+
+1. **Update systems:** these systems run each schedule iteration if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(apply_damage.in_state(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(generate_map.on_enter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified state is exited.
+   1. `app.add_system(autosave.on_exit(GameState::Playing))`
+
+Update systems are by far the simplest: they're simply powered by run criteria.
+`.in_state` is precisely equivalent to `run_if_resource_equals`, except with an additional trait bound that the resource must implement `State`.
+
+On-enter and on-exit systems are stored in dedicated schedules, two per state, within the `App's` `Schedules`.
+These schedules can be configured in all of the ordinary ways, but, as they live in different schedules, ordering cannot be defined relative to systems in the main schedule.
+
+You can change which variant of a state you are in by requesting the state as a mutable resource, then using the `State::transition_to` method within your system.
+Due to their disruptive and far-reaching effects, state transitions do not occur immediately.
+Instead, they are deferred (like commands), until the next `flush_state<S: State>` exclusive system runs.
+This system first runs the `on_exit` schedule of the previous state on the world, then runs the `on_enter` schedule of the new state on the world.

```

Comment:
Do commands run between exit and enter?

Reply:
Flushing commands in between state transitions would be valid, but I think we should leave it to the user to flush in their exit/enter schedules.

I'm leaning toward not having any intrinsic flushes.